libp2p/NodeTable.cpp

```diff
@@ -621,36 +630,56 @@ void NodeTable::onPacketReceived(
 
 void NodeTable::doDiscovery()
 {
-    m_timers.schedule(c_bucketRefresh.count(), [this](boost::system::error_code const& _ec)
-    {
-        if (_ec)
-            // we can't use m_logger here, because captured this might be already destroyed
+    m_discoveryTimer->expires_from_now(c_bucketRefreshMs);
+    auto discoveryTimer{m_discoveryTimer};
+    m_discoveryTimer->async_wait([this, discoveryTimer](boost::system::error_code const& _ec) {
+        // We can't use m_logger if an error occurred because captured this might be already
+        // destroyed
+        if (_ec.value() == boost::asio::error::operation_aborted ||
+            discoveryTimer->expires_at() == c_steadyClockMin)
+        {
+            clog(VerbosityDebug, "discov") << "Discovery timer was probably cancelled";
+            return;
+        }
+        else if (_ec)
+        {
             clog(VerbosityDebug, "discov")
-                << "Discovery timer was probably cancelled: " << _ec.value() << " "
-                << _ec.message();
-
-        if (_ec.value() == boost::asio::error::operation_aborted || m_timers.isStopped())
+                << "Discovery timer error detected: " << _ec.value() << " " << _ec.message();
             return;
+        }
 
         NodeID randNodeId;
         crypto::Nonce::get().ref().copyTo(randNodeId.ref().cropped(0, h256::size));
         crypto::Nonce::get().ref().copyTo(randNodeId.ref().cropped(h256::size, h256::size));
-        LOG(m_logger) << "Random discovery: " << randNodeId;
-        doDiscover(randNodeId, 0, make_shared<set<shared_ptr<NodeEntry>>>());
+        LOG(m_logger) << "Starting discovery algorithm run for random node id: " << randNodeId;
+        doDiscoveryRound(randNodeId, 0 /* round */, make_shared<set<shared_ptr<NodeEntry>>>());
     });
 }
 
-void NodeTable::doHandleTimeouts()
+void NodeTable::doProcessEvictions()
```

Comment:
Not sure about the naming, because it handles not only evictions from the buckets, but also just initial Pings timed out. I'd leave it with "timeouts" in the name

Reply:
Ah, good point, I didn't realize that.